### ISSUE TITLE
incompatible_msrv false negative: tests

### ISSUE THREAD

**Issue by bryceberger** [rust-lang/rust-clippy#14277]
### Summary

I disagree with the resolution of #12257, and think that `incompatible_msrv` **should** apply to tests. Given the conflicting desires, I think it would make sense to provide a config option for this lint.

I would like something akin to the following:
```toml
# clippy.toml
check-incompatible-msrv-in-tests = true
```

Background: [jj](https://github.com/jj-vcs/jj/) currently has an MSRV of `1.76`. However, it's encouraged to use newer / nightly compilers during development, to take advantage of speed improvements / parallel frontend / etc. As such, they do not force a version using `rust-toolchain.toml`. A PRs that added a `rust-toolchain.toml` pin was closed under the assumption that `incompatible_msrv` would help with this (https://github.com/jj-vcs/jj/pull/1913#issuecomment-2037636332).

#12257 makes the argument:

> However, test-only code does not prevent the crate from being used with the specified MSRV, as a library. Tests are not compiled when a crate is used as a dependency.

However, tests are compiled when the tests are ran in CI. We would like to keep the CI pinned to MSRV, and do not want to have separate `Build (MSRV)` and `Test (latest stable)` CI jobs.

### Lint Name

incompatible_msrv

### Reproducer

```toml
# Cargo.toml
[package]
name = "repro"
edition = "2021"
rust-version = "1.76"
```

```rust
// src/lib.rs
#[test]
fn repro() {
    assert!(std::fs::exists("/").unwrap());
}
```

```bash
$ cargo clippy --tests
# no output...
```

### Version

```text
rustc 1.86.0-nightly (854f22563 2025-01-31)
binary: rustc
commit-hash: 854f22563c8daf92709fae18ee6aed52953835cd
commit-date: 2025-01-31
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.7
```
